### PR TITLE
Stop allocating vtable entries for non-object-safe methods

### PR DESCRIPTION
**Pull request by nbdd0121**
Current a vtable entry is allocated for all associated fns, even if the method is not object-safe: https://godbolt.org/z/h7vx6f35T

As a result, each vtable for `Iterator`' currently consumes 74 `usize`s. This PR stops allocating vtable entries for those methods, reducing vtable size of each `Iterator` vtable to 7 `usize`s.

Note that this PR introduces will cause more invocations of `is_vtable_safe_method`. So a perf run might be needed. If result isn't favorable then we might need to query-ify `is_vtable_safe_method`.

### CODE REVIEW

**Comment by nbdd0121** [comment]
Introduced a query. Please do another perf run

File: compiler/rustc_trait_selection/src/traits/mod.rs [hunk 1]
```diff
@@ -641,21 +666,15 @@ fn vtable_entries<'tcx>(
...
-                // Now list each method's DefId and InternalSubsts (for within its trait).
-                // If the method can never be called from this object, produce `Vacant`.
-                let own_entries = trait_methods.map(move |trait_method| {
-                    debug!("vtable_entries: trait_method={:?}", trait_method);
-                    let def_id = trait_method.def_id;
-
-                    // Some methods cannot be called on an object; skip those.
-                    if !is_vtable_safe_method(tcx, trait_ref.def_id(), &trait_method) {
-                        debug!("vtable_entries: not vtable safe");
-                        return VtblEntry::Vacant;
```
Comment (nagisa) [inline compiler/rustc_trait_selection/src/traits/mod.rs:657]: Is there any reason why we cannot remove the vtable entries for the "impossible predicates" situation below and remove the `Vacant` variant entirely?

Doesn't need to be considered for this PR, of course, just a passing thought.

**Comment by nagisa** [comment]
I would like to see additional tests for this, especially good would be a test for `Iterator` (to verify the claims made in the PR description) as well as a test for a vtable where we still have `Vacant` entries of some sort due to the `impossible_predicates` situation.

File: compiler/rustc_trait_selection/src/traits/mod.rs (hunk 1 above)
Comment (nbdd0121) [reply compiler/rustc_trait_selection/src/traits/mod.rs:657]:
```rust
trait X {
    fn foo(&self)
    where
        Self: Send,
    {
    }
}

impl X for () {
    fn foo(&self) {}
}

impl X for *mut () {}

pub fn main() {
    <dyn X + Send as X>::foo(&());
}
```

We need to generate `X::foo` for `()`, but we cannot generate `X::foo` for `*mut ()`.

**Comment by nagisa** [comment]
@bors r+

Thanks!

**Comment by camelid** [comment]
@bors r-

CI is failing, looks like it just needs a quick `x.py test --bless`.

**Comment by nbdd0121** [comment]
Done (I manually edited the file so I don't have to recompile everything)

**Comment by nbdd0121** [comment]
Race condition :)

@bors r=nagisa

**Comment by camelid** [comment]
@bors r-

CI is failing again; please wait for CI to pass before re-approving the PR (otherwise it slows down the bors queue) :)

**Comment by camelid** [comment]
Heh, another race condition :)

**Comment by Mark-Simulacrum** [comment]
This had [mixed](https://perf.rust-lang.org/compare.html?calcNewSig=true&start=e2750baf53aaa60db95f10759f6cf9463dc5a6bd&end=e30b68353fe22b00f40d021e7914eeb78473b3c1) results on landing, generally, with no significant cycle count results. Marking as triaged.

**Pull request merged**